content/news/2022-07-29-bevy-0.8/index.md

```diff
@@ -0,0 +1,1131 @@
+
+This "flight helmet" scene consists of many "pieces" nested under the main helmet entity. Hiding all of these "sub entities" is as now as easy as hiding the top level helmet entity.
+
+<video controls loop><source  src="visibility_inheritance.mp4" type="video/mp4"/></video>
+
+```rust
+fn hide_helmets(mut helmet_visibilities: Query<&mut Visibility, With<Helmet>>) {
+    let mut helmet_visibility = helmet_visibilities.single_mut();
+    helmet_visibility.is_visible = false;
+}
+```
+
+In past versions of Bevy, you had to hide each piece manually!
+
+The "inherited visibility" is computed in the [`PostUpdate`] stage and stored on the [`ComputedVisibility`] component. [`ComputedVisibility`] now has the following functions:
+* `is_visible_in_hierarchy()`: Whether or not the entity is visible according to "visibility inheritance".
+* `is_visible_in_view()`: Whether or not the entity is visible in any view. This is used for culling entities in cases like "frustum culling".
+* `is_visible()`: The canonical place to determine whether or not an entity will be drawn. Combines "view visibility" and "hierarchy visibility".
+
+[`PostUpdate`]: https://docs.rs/bevy/0.8.0/bevy/app/enum.CoreStage.html#variant.PostUpdate
+[`Visibility`]: https://docs.rs/bevy/0.8.0/bevy/render/view/struct.Visibility.html
+[`ComputedVisibility`]: https://docs.rs/bevy/0.8.0/bevy/render/view/struct.ComputedVisibility.html
+
+
+## Built-in Shader Modularization
+
+<div class="release-feature-authors">authors: Rob Swain (@superdump)</div>
+
+In **Bevy 0.8**, we've (started) modularizing our built-in shaders. Notably, this means you can now import and run the built in PBR shader / lighting logic:
+
+```rust
+#import bevy_pbr::utils
+#import bevy_pbr::clustered_forward
+#import bevy_pbr::lighting
+#import bevy_pbr::shadows
+#import bevy_pbr::pbr_types
+#import bevy_pbr::pbr_functions
+
+@fragment
+fn fragment(in: FragmentInput) -> @location(0) vec4<f32> {
+    var pbr_input: PbrInput = pbr_input_new();
+    // set the base color to red
+    pbr_input.material.base_color = vec4<f32>(1.0, 0.0, 0.0, 1.0);
+    /* set other PbrInput fields here */
+    return tone_mapping(pbr(pbr_input))
+}
+```
+
+We've also modularized our mesh and view binding shader logic. When paired with the [New Material System](#new-material-system), users can now reasonably write custom PBR Materials without re-defining all of the PBR logic in their shaders. The new [Array Texture Example](https://github.com/bevyengine/bevy/blob/v0.8.0/examples/shader/array_texture.rs) illustrates how to define a custom PBR shader Material using these new APIs:
+
+![array texture](array_texture.png)
+
+We also plan on evolving the user experience here. Now that we've broken everything up and modularized it, we'll work on reducing the amount of boilerplate required to extend this logic (cutting down on imports, removing the need to set all PbrInput fields, etc).  
```

Comment:
Sounds like you have ideas that I’d like to hear about. :)

Reply:
I (think) we've talked about most of them already. I don't have strong opinions yet, other than that this is still too boilerplatey / tied to internal implementation details (such as how each import is organized and ordered). Things in the vein of:

1. templates that handle bindings automatically (ex: nest the user's fragment() function inside of the actual fragment function and handle bindings + pbr setup automatically "godot style").
2. merge imports into a single `#import bevy_pbr::custom_pbr_shader` line (maybe in combination with import de-duping logic).
3. pre-calculating things using the `var foo: vec4<f32>` / `compute_foo()` foo pattern I outlined in this set of prs.